rust/src/operations/restore.rs

```diff
@@ -0,0 +1,288 @@
+    /// If true, continue to run when encountering missing files.
+    pub fn ignore_missing_files(mut self, ignore_missing_files: bool) -> Self {
+        self.ignore_missing_files = ignore_missing_files;
+        self
+    }
+
+    /// Set whether allow to downgrade protocol
+    pub fn protocol_downgrade_allowed(mut self, protocol_downgrade_allowed: bool) -> Self {
+        self.protocol_downgrade_allowed = protocol_downgrade_allowed;
+        self
+    }
+}
+
+async fn execute(
+    object_store: ObjectStoreRef,
+    snapshot: DeltaTableState,
+    version_to_restore: Option<i64>,
+    datetime_to_restore: Option<DateTime<Utc>>,
+    ignore_missing_files: bool,
+    protocol_downgrade_allowed: bool,
+) -> DeltaResult<RestoreMetrics> {
+    if !(version_to_restore
+        .is_none()
+        .bitxor(datetime_to_restore.is_none()))
+    {
+        return Err(DeltaTableError::from(RestoreError::InvalidRestoreParameter));
+    }
+    let mut table = DeltaTable::new(object_store.clone(), DeltaTableConfig::default());
+    let version = match datetime_to_restore {
+        Some(datetime) => {
+            table.load_with_datetime(datetime).await?;
+            table.version()
+        }
+        None => {
+            table.load_version(version_to_restore.unwrap()).await?;
+            table.version()
+        }
+    };
+
+    if version >= snapshot.version() {
+        return Err(DeltaTableError::from(RestoreError::TooLargeRestoreVersion(
+            version,
+            snapshot.version(),
+        )));
+    }
+    let state_to_restore_files = table.get_state().files().clone();
+    let latest_state_files = snapshot.files().clone();
+    let state_to_restore_files_set =
+        HashSet::<Add>::from_iter(state_to_restore_files.iter().cloned());
+    let latest_state_files_set = HashSet::<Add>::from_iter(latest_state_files.iter().cloned());
+
+    let files_to_add: Vec<Add> = state_to_restore_files
+        .into_iter()
+        .filter(|a: &Add| !latest_state_files_set.contains(a))
+        .map(|mut a: Add| -> Add {
+            a.data_change = true;
+            a
+        })
+        .collect();
+
+    let deletion_timestamp = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .unwrap()
+        .as_millis() as i64;
+    let files_to_remove: Vec<Remove> = latest_state_files
+        .into_iter()
+        .filter(|a: &Add| !state_to_restore_files_set.contains(a))
+        .map(|a: Add| -> Remove {
+            Remove {
+                path: a.path.clone(),
+                deletion_timestamp: Some(deletion_timestamp),
+                data_change: true,
+                extended_file_metadata: Some(false),
+                partition_values: Some(a.partition_values.clone()),
+                size: Some(a.size),
+                tags: a.tags,
+            }
+        })
+        .collect();
+
+    if !ignore_missing_files {
+        check_files_available(object_store.as_ref(), &files_to_add).await?;
+    }
```

Comment:
interesting - would not have thought that it is useful, to delivberately create a state with missing files, but it seems the reference implementation also supports this :).

Reply:
The data file may missing for many reason, deleted unintentionally or by vacuum operation.This will cause data missing if always allow to restore, so we must make users aware of the problem.